waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -1704,6 +1738,170 @@
+                        object-id->timer (pc/map-vals (fn [_] (timers/start (metrics/waiter-timer "scheduler" scheduler-name "query-events"))) object-id->chan)]
+                    ;; update state for all cached events
+                    (doseq [[object-id events] object-id->events]
+                      (log/info "updating state with cached events for" object-id events)
+                      (swap! watch-state
+                             #(as-> % state
+                                ((object-id->update-in-state-fn object-id) state events)))
+                      (let [events-update (merge events {:object-id object-id :time (t/now)})]
+                        (async/put! fetch-events-chan events-update)))
+                    ;; take all responses from channels, processing them as appropriate
+                    (loop [event-chan->object-id (cljset/map-invert object-id->chan)]
+                      (when (pos? (count event-chan->object-id))
+                        (let [event-chans (keys event-chan->object-id)
+                              [response-or-exc current-chan] (async/alts! event-chans)
+                              object-id (get event-chan->object-id current-chan)
+                              object-key (if (= unhealthy-service-id object-id)
+                                           (get-service-key unhealthy-service)
+                                           (get-pod-key (get pod-id->unhealthy-pod object-id)))]
+                          (timers/stop (get object-id->timer object-id))
+                          (if (instance? Throwable response-or-exc)
+                            (let [{:keys [http-utils/response]} (ex-data response-or-exc)
+                                  exc (or response
+                                          response-or-exc)]
+                              (log/info "Failed to acquire events for object" object-key exc)
+                              (cu/cache-evict k8s-object-key->event-cache object-key))
+                            (let [events (extract-events-from-k8s-response response-or-exc)]
+                              (cu/cache-put! k8s-object-key->event-cache object-key events)
+                              (swap! watch-state
+                                     #(as-> % state
+                                        ((object-id->update-in-state-fn object-id) state events)))
+                              (let [events-update (merge events {:object-id object-id :time (t/now)})]
+                                (async/put! fetch-events-chan events-update))))
+                          (recur (dissoc event-chan->object-id current-chan))))))))))
+          (utils/sleep 10000)
+          (recur (inc iter))))
+      (catch Throwable t
+        (when exit-on-error?
+          (log/error t "unrecoverable error in K8s event fetcher thread, terminating waiter.")
+          (System/exit 1))))))
```

Comment:
Do we want to exit or allow degraded operation? We can have health check on update time metrics from this block to known when this daemons has stopped running in a timely fashion.

Reply:
I agree with this sentiment; this feature is an added-value and should not adversely impact core functionality.